### ISSUE TITLE
FiPy coupled PDEs help

### ISSUE THREAD

**Issue by WenjiangHuang111** [usnistgov/fipy#758]
Hi,
I have a coupled PDEs that I wanna use FiPY to solve it but I am confused on defining the equations and boundary conditions. The description of the PDEs show in the PDF file below.
[Coupled PDEs.pdf](https://github.com/usnistgov/fipy/files/5371643/Coupled.PDEs.pdf)

Thanks all for any suggestion and help.
My Python code is shown as below:

```python
import math
import numpy as np
from fipy import *
import matplotlib.pyplot as plt
k = 8.617333262145e-5 # Boltzmann_Constant, eV*K
E_i_f = 4.0           # eV
E_v_f = 1.4 
S_v_f = 5*k
T = 500               # K
C_i_t0 = math.exp(-E_i_f/(k*T))
C_v_t0 = math.exp(S_v_f/k)*math.exp(-E_v_f/(k*T))
d_Fev0 = 1.02e-4     # m^2/s
d_Crv0 = 2.3e-4 
d_Fei0 = 6.59e-7 
d_Cri0 = 1.25e-7
E_Fev = 1.2          # eV
E_Crv = 1.2 
E_Fei = 0.3
E_Cri = 0.9
d_Fev = d_Fev0*math.exp(-E_Fev/(k*T))
d_Crv = d_Crv0*math.exp(-E_Crv/(k*T))
d_Fei = d_Fei0 * math.exp(-E_Fei/(k*T))
d_Cri = d_Cri0 * math.exp(-E_Cri/(k*T))
Omega = 1.18e-29 # m^3
d_rec = 5.7e-10 # m
ps =1e12 # m^-2
Ko = 1e-6
zv = 1.0
zj = 1.2

L = 3e-8
nx = 1000
dx = L/nx
mesh = Grid1D(nx = nx, dx = dx)
x = mesh.cellCenters[0]
C_Fe_t0 = 0.84
C_Cr_t0 = 0.16
u1 = CellVariable(name='u1',mesh=mesh,value=C_v_t0, hasOld=True)
u2 = CellVariable(name='u2',mesh=mesh,value=C_i_t0, hasOld=True)
u3 = CellVariable(name='u3',mesh=mesh,value=C_Fe_t0, hasOld=True)
u4 = CellVariable(name='u4',mesh=mesh,value=C_Cr_t0, hasOld=True)

u1.constrain(C_v_t0, mesh.facesLeft)
u2.constrain(C_i_t0, mesh.facesRight)
u3.faceGrad.constrain(0,mesh.facesLeft)
u4.faceGrad.constrain(0,mesh.facesLeft)
u1.faceGrad.constrain(0,mesh.facesRight)
u2.faceGrad.constrain(0,mesh.facesRight)
u3.faceGrad.constrain(0,mesh.facesRight)
u4.faceGrad.constrain(0,mesh.facesRight)

RCvCi = (4*np.pi*d_rec/Omega)*(ImplicitSourceTerm(coeff=(d_Fev+d_Fei)*u1*u2, var = u3)+
            ImplicitSourceTerm(coeff=(d_Crv+d_Cri)*u1*u2,var=u4))

# eq1 = (TransientTerm(var=u1) == -(ConvectionTerm(coeff=(d_Fev-d_Crv)*u3.faceGrad, var = u1)
#     - ConvectionTerm(coeff=d_Fev*u1.faceGrad, var = u3) - ConvectionTerm(coeff=d_Crv*u1.faceGrad, var = u4))
#     + Ko - RCvCi)

# eq2 = (TransientTerm(var=u2) == -(ConvectionTerm(coeff=(d_Cri-d_Fei)*u3.faceGrad, var = u2)
#     - ConvectionTerm(coeff=d_Fei*u2.faceGrad, var = u3) - ConvectionTerm(coeff=d_Cri*u2.faceGrad, var = u4))
#     + Ko - RCvCi)

# eq3 = (TransientTerm(var=u3) == ConvectionTerm(coeff=d_Fei*u2.faceGrad,var = u3) 
#                                 + ConvectionTerm(coeff=d_Fei*u3.faceGrad,var = u2)
#                                 - ConvectionTerm(coeff=d_Fev*u1.faceGrad,var = u3)
#                                 + ConvectionTerm(coeff=d_Fev*u3.faceGrad,var = u1))

# eq4 = (TransientTerm(var=u4) == ConvectionTerm(coeff=d_Cri*u2.faceGrad,var = u4) 
#                                 + ConvectionTerm(coeff=d_Cri*u4.faceGrad,var = u2)
#                                 - ConvectionTerm(coeff=d_Crv*u1.faceGrad,var = u4)
#                                 + ConvectionTerm(coeff=d_Crv*u4.faceGrad,var = u1))


# u1 = C_v, u2 = C_i, u3 = C_Fe (or C_A), u3 = C_Cr (or C_B)
eq1 = (TransientTerm(var=u1) == DiffusionTerm(coeff = u1*(d_Crv-d_Fev), var = u3) 
                              + DiffusionTerm(coeff = d_Crv*u4+d_Fev*u3, var = u1)+ Ko - RCvCi)

eq2 = (TransientTerm(var=u2)==DiffusionTerm(coeff = u2*(d_Fei-d_Cri), var = u3) 
        + DiffusionTerm(coeff = d_Cri*u4+d_Fei*u3, var = u2)
        + ConvectionTerm(coeff=2*(d_Fei-d_Cri)*u3.faceGrad, var = u2)
            -DiffusionTerm(coeff=2*(d_Fei-d_Cri)*u2,var=u3)+ Ko-RCvCi)


eq3 = (TransientTerm(var=u3)==DiffusionTerm(coeff = u2*d_Fei + u1*d_Fev, var = u3) 
       + DiffusionTerm(coeff = d_Fei*u3, var = u2)-DiffusionTerm(coeff = d_Fev*u3, var = u1)
       + ConvectionTerm(coeff=2*d_Fei*u3.faceGrad, var = u2)
       - DiffusionTerm(coeff = 2*d_Fei*u2, var = u3))

eq4 = (TransientTerm(var=u4)==DiffusionTerm(coeff = u2*d_Cri + u1*d_Crv, var = u4) 
       + DiffusionTerm(coeff = d_Cri*u4, var = u2)-DiffusionTerm(coeff = d_Crv*u4, var = u1)
       + ConvectionTerm(coeff=2*d_Cri*u4.faceGrad, var = u2)
       - DiffusionTerm(coeff = 2*d_Cri*u2, var = u4))

eq = eq1 & eq2 & eq3 & eq4
plt.figure(figsize=(14,8))
viewer1=Matplotlib1DViewer(vars=u1, axes=plt.subplot(1, 4, 1),title='$C_{v}$')
viewer2=Matplotlib1DViewer(vars=u2, axes=plt.subplot(1, 4, 2),title='$C_{i}$')
viewer3=Matplotlib1DViewer(vars=u3, axes=plt.subplot(1, 4, 3),title='$C_{Fe}$',datamin=0, datamax=1)
viewer4=Matplotlib1DViewer(vars=u4, axes=plt.subplot(1, 4, 4),title='$C_{Cr}$',datamin=0, datamax=1)
# viewer1 = MatplotlibViewer(vars = u1)
# viewer2 = MatplotlibViewer(vars = u2)
# viewer3 = MatplotlibViewer(vars = u3)
# viewer4 = MatplotlibViewer(vars = u4)
plt.tight_layout()

dt = 1e-5
for i in range(1000):
    u1.updateOld()
    u2.updateOld()
    u3.updateOld()
    u4.updateOld()   
    res=eq.sweep(dt=1) # a large time step dt=5 has been chosen because there is no transient term actually present in my pde.
    if(i%100==0):
#         plt.figure(figsize=(10,8))
        viewer1.plot()
        viewer2.plot()
        viewer3.plot()
        viewer4.plot()

```

**Comment by guyer** [comment]
Your equation definitions look reasonable, although:
- I would convert `ConvectionTerm` to `DiffusionTerm` as it's more stable, e.g., `ConvectionTerm(coeff=2*(d_Fei-d_Cri)*u3.faceGrad, var = u2)` becomes `ConvectionTerm(coeff=2*(d_Fei-d_Cri)*u2, var = u3)`
- Using completely different symbols in your code than in your PDF makes it really hard to follow. I got lost in all the `u1`'s and `C_Fe_t0`'s, since none of that is shown in your math.

As to boundary conditions, 
- FiPy is no-flux by default, so remove all of the `ui.faceGrad.constrain()` statements.
- $\partial C_A(t,1)/\partial t = 0$ looks equivalent to $C_A(t,1) = C_A(t=0)$, a Dirichlet condition.

**Comment by WenjiangHuang111** [comment]
> Your equation definitions look reasonable, although:
> 
> * I would convert `ConvectionTerm` to `DiffusionTerm` as it's more stable, e.g., `ConvectionTerm(coeff=2*(d_Fei-d_Cri)*u3.faceGrad, var = u2)` becomes `ConvectionTerm(coeff=2*(d_Fei-d_Cri)*u2, var = u3)`
> * Using completely different symbols in your code than in your PDF makes it really hard to follow. I got lost in all the `u1`'s and `C_Fe_t0`'s, since none of that is shown in your math.
> 
> As to boundary conditions,
> 
> * FiPy is no-flux by default, so remove all of the `ui.faceGrad.constrain()` statements.
> * $\partial C_A(t,1)/\partial t = 0$ looks equivalent to $C_A(t,1) = C_A(t=0)$, a Dirichlet condition.

I am very sorry for typo, refer to the new PDF file I just revise .
[Coupled PDEs.pdf](https://github.com/usnistgov/fipy/files/5372255/Coupled.PDEs.pdf)

In Eq2 (refer to Eq 10 in PDF), there is a term 2*(d_Fei-d_Cri)*div Ci*div CFe. I was wondering how to express it using convection term. That is why I used  `ConvectionTerm(coeff=2*(d_Fei-d_Cri)*u3.faceGrad, var = u2)  - DiffusionTerm(coeff = 2*d_Fei*u2, var = u3)` I am not sure if it is correct. 

for the BC right, \partial Fe/ \partial t = 0 is not the same as its concentration equal to 0. In face, the concentration in BC right of Fe should be constant (as initial value 0.84. same for Cr)

u1-> Cv, u2->Ci, u3->C_Fe, u4->C_Cr

Sorry again for any confusion.

WJ

**Comment by guyer** [comment]
Don't run the chain rule all the way through as you have in (9)-(12).

As discussed in [Discretization](https://www.ctcms.nist.gov/fipy/documentation/numerical/discret.html#discretization), FiPy wants terms that it can apply the divergence theorem to, like <img src="https://render.githubusercontent.com/render/math?math=\nabla\cdot(\vec{u}\phi)"> and <img src="https://render.githubusercontent.com/render/math?math=\nabla\cdot(\Gamma\nabla\phi)">. Given a choice, the latter (diffusive) is preferable to the former (convective), as it is numerically more stable. The divergence theorem cannot be applied to expressions like <img src="https://render.githubusercontent.com/render/math?math=\phi\nabla^2 \psi"> or <img src="https://render.githubusercontent.com/render/math?math=\nabla \phi\cdot\nabla \psi">, so, when you have these terms, you should use the chain rule to find a more favorable form.

Fortunately, in your case, it's easy. 
<img src="https://render.githubusercontent.com/render/math?math=%5Cbegin%7Balign*%7D%0A%5Cfrac%7B%5Cpartial%20C_v%7D%7B%5Cpartial%20t%7D%20%26%3D%20-%5Cnabla%5Ccdot%20J_v%20%2B%20%5Ccdots%0A%5C%5C%0A%26%3D%20-%5Cnabla%5Ccdot%5Cleft%5B%20(d_%7BFe%20v%7D%20-%20d_%7BCr%20v%7D)%20C_v%20%5Cnabla%20C_%7BFe%7D%5Cright%5D%20-%20%5Cnabla%5Ccdot%5Cleft%5B%5Ccdots%20%5Cright%20%5D%20%2B%20%5Ccdots%0A%5Cend%7Balign*%7D">
is in the form that FiPy wants. 

I do strongly recommend that you take more care with your notation. Since <img src="https://render.githubusercontent.com/render/math?math=J_v%20%3D%20(d_%7BFe%20v%7D%20-%20d_%7BCr%20v%7D)%20C_v%20%5Cnabla%20C_%7BFe%7D%20%2B%20%5Ccdots"> is a flux (rank-1), <img src="https://render.githubusercontent.com/render/math?math=%5Cnabla%20J_v"> would be a tensor (rank-2). I realize that there doesn't appear to be any difference in 1D, but based on sixteen years of trying to help people with questions like this, I have come to the conclusion that everything that anybody has ever derived in 1D is wrong.

As to BCs, I didn't say concentration equal to zero. I said concentration equal to concentration at t=0.

**Comment by WenjiangHuang111** [comment]
> Don't run the chain rule all the way through as you have in (9)-(12).
> 
> As discussed in [Discretization](https://www.ctcms.nist.gov/fipy/documentation/numerical/discret.html#discretization), FiPy wants terms that it can apply the divergence theorem to, like ![](https://render.githubusercontent.com/render/math?math=\nabla\cdot(\vec{u}\phi)) and ![](https://render.githubusercontent.com/render/math?math=\nabla\cdot(\Gamma\nabla\phi)). Given a choice, the latter (diffusive) is preferable to the former (convective), as it is numerically more stable. The divergence theorem cannot be applied to expressions like ![](https://render.githubusercontent.com/render/math?math=\phi\nabla^2 \psi) or ![](https://render.githubusercontent.com/render/math?math=\nabla \phi\cdot\nabla \psi), so, when you have these terms, you should use the chain rule to find a more favorable form.
> 
> Fortunately, in your case, it's easy.
> ![](https://render.githubusercontent.com/render/math?math=%5Cbegin%7Balign*%7D%0A%5Cfrac%7B%5Cpartial%20C_v%7D%7B%5Cpartial%20t%7D%20%26%3D%20-%5Cnabla%5Ccdot%20J_v%20%2B%20%5Ccdots%0A%5C%5C%0A%26%3D%20-%5Cnabla%5Ccdot%5Cleft%5B%20(d_%7BFe%20v%7D%20-%20d_%7BCr%20v%7D)%20C_v%20%5Cnabla%20C_%7BFe%7D%5Cright%5D%20-%20%5Cnabla%5Ccdot%5Cleft%5B%5Ccdots%20%5Cright%20%5D%20%2B%20%5Ccdots%0A%5Cend%7Balign*%7D)
> is in the form that FiPy wants.
> 
> I do strongly recommend that you take more care with your notation. Since ![](https://render.githubusercontent.com/render/math?math=J_v%20%3D%20(d_%7BFe%20v%7D%20-%20d_%7BCr%20v%7D)%20C_v%20%5Cnabla%20C_%7BFe%7D%20%2B%20%5Ccdots) is a flux (rank-1), ![](https://render.githubusercontent.com/render/math?math=%5Cnabla%20J_v) would be a tensor (rank-2). I realize that there doesn't appear to be any difference in 1D, but based on sixteen years of trying to help people with questions like this, I have come to the conclusion that everything that anybody has ever derived in 1D is wrong.
> 
> As to BCs, I didn't say concentration equal to zero. I said concentration equal to concentration at t=0.

Sorry for misunderstanding your explanation for BCs.

if I got your idea correctly, the eqs should be modified in this way? (I renamed the variables for easily read)
```
C_v = CellVariable(name='C_v', mesh=mesh, value=C_v_t0, hasOld=True)
C_i = CellVariable(name='C_i', mesh=mesh, value=C_i_t0, hasOld=True)
C_Fe = CellVariable(name='C_Fe', mesh=mesh, value=C_Fe_t0, hasOld=True)
C_Cr = CellVariable(name='C_Cr', mesh=mesh, value=C_Cr_t0, hasOld=True)

C_v.constrain(C_v_t0, mesh.facesLeft)
C_i.constrain(C_i_t0, mesh.facesRight)

RCvCi = (4*np.pi*d_rec/Omega)*(ImplicitSourceTerm(coeff=(d_Fev+d_Fei)*C_v*C_i, var = C_Fe)+
            ImplicitSourceTerm(coeff=(d_Crv+d_Cri)*C_v*C_i,var=C_Cr))

eq1 = (TransientTerm(var=C_v) == - DiffusionTerm(coeff=(d_Fev-d_Crv)*C_v, var = C_Fe)
                               + DiffusionTerm(coeff=d_Fev*C_Fe+d_Crv*C_Cr, var = C_v) + Ko - RCvCi)

eq2 = (TransientTerm(var=C_i) ==  DiffusionTerm(coeff= (d_Fei-d_Cri)*C_i, var = C_Fe)
                               + DiffusionTerm(coeff= d_Fei*C_Fe+d_Cri*C_Cr, var = C_i) + Ko - RCvCi)

eq3 = (TransientTerm(var=C_Fe) == DiffusionTerm(coeff=d_Fei*C_Fe,var = C_i) 
                              + DiffusionTerm(coeff=d_Fei*C_i,var = C_Fe)
                              - DiffusionTerm(coeff=d_Fev*C_Fe,var = C_v)
                              + DiffusionTerm(coeff=d_Fev*C_v,var = C_Fe))

eq4 = (TransientTerm(var=C_Cr) == DiffusionTerm(coeff=d_Cri*C_Cr,var = C_i) 
                              + DiffusionTerm(coeff=d_Cri*C_i,var = C_Cr)
                              - DiffusionTerm(coeff=d_Crv*C_Cr,var = C_v)
                              + DiffusionTerm(coeff=d_Crv*C_v,var = C_Cr))
```

The curve I got looks unreasonable to me. Did I mess up something here?

For BCs,  are "Flux of Fe is zero at x=0" and "partial Fe/ partial t =0 at x=1" the same?  Cz for 1D, as you said, Jv and div Jv don't appear any difference. 

Thanks for your suggestions for these notations. I currently want to reproduce the result from some works and then I just use the PDEs they claimed.

**Comment by guyer** [comment]
> The curve I got looks unreasonable to me. Did I mess up something here?

The equations look OK to me.

For the boundary conditions, you constrain `C_v` on the left, but `C_i` on the right:
```python
C_v.constrain(C_v_t0, mesh.facesLeft)
C_i.constrain(C_i_t0, mesh.facesRight)
```

> For BCs, are "Flux of Fe is zero at x=0" and "partial Fe/ partial t =0 at x=1" the same? Cz for 1D, as you said, Jv and div Jv don't appear any difference.

No. Saying that <img src="https://render.githubusercontent.com/render/math?math=%5Cfrac%7B%5Cpartial%20C_%7BFe%7D%7D%7B%5Cpartial%20t%7D%20%3D%200"> at x=1 is equivalent to saying that <img src="https://render.githubusercontent.com/render/math?math=%5Cnabla%5Ccdot%20J_%7BFe%7D%20%3D%200"> at x=1. In other words, the flux of Fe at <img src="https://render.githubusercontent.com/render/math?math=x%3D1-%5Cvarepsilon"> is the same as at <img src="https://render.githubusercontent.com/render/math?math=x%3D1%2B%5Cvarepsilon">. It doesn't say anything about the *value* of the flux, though.

**Comment by WenjiangHuang111** [comment]
> > The curve I got looks unreasonable to me. Did I mess up something here?
> 
> The equations look OK to me.
> 
> For the boundary conditions, you constrain `C_v` on the left, but `C_i` on the right:
> 
> ```python
> C_v.constrain(C_v_t0, mesh.facesLeft)
> C_i.constrain(C_i_t0, mesh.facesRight)
> ```
> 
> > For BCs, are "Flux of Fe is zero at x=0" and "partial Fe/ partial t =0 at x=1" the same? Cz for 1D, as you said, Jv and div Jv don't appear any difference.
> 
> No. Saying that ![](https://render.githubusercontent.com/render/math?math=%5Cfrac%7B%5Cpartial%20C_%7BFe%7D%7D%7B%5Cpartial%20t%7D%20%3D%200) at x=1 is equivalent to saying that ![](https://render.githubusercontent.com/render/math?math=%5Cnabla%5Ccdot%20J_%7BFe%7D%20%3D%200) at x=1. In other words, the flux of Fe at ![](https://render.githubusercontent.com/render/math?math=x%3D1-%5Cvarepsilon) is the same as at ![](https://render.githubusercontent.com/render/math?math=x%3D1%2B%5Cvarepsilon). It doesn't say anything about the _value_ of the flux, though. Because there are no sources or sinks of Fe, that means that C_Fe is constant.
> 
> Saying that the flux of Fe is zero at x=0 says nothing about how the value of the flux is changing at x=0, thus the divergence can be non-zero, leading to accumulation or depletion at that boundary. This is a Neumann condition. More importantly, no-flux is the natural boundary condition for finite volume and so is addressed automatically by FiPy unless you apply some other constraint.

Thanks for correcting the BCs. Then I am wondering if the current BCs in my code are enough? for x=0, J_Fe and J_Cr are zero and they don't need to be specified because FiPy treats it as default. For x=1, Neumann condition for Fe and Cr, should I add

```
C_Fe.faceGrad.constrain(0, mesh.facesRight)
C_Cr.faceGrad.constrain(0, mesh.facesRight) 
```
?
It doesn't change much to the result. At the left, x=0, I should observe the change of Fe and Cr concentration (enrichment and depletion, respectively) and at x=1, they should be same as their initial values, as you said, C(t=0). However, I can only observe two horizontal lines for Fe and Cr.

**Comment by guyer** [comment]
You don't have a Neumann condition for Fe and Cr at x=1. In addition to <img src="https://render.githubusercontent.com/render/math?math=J_%7BFe%7D(t%2C%200)%20%3D%20J_%7BCr%7D(t%2C%200)%20%3D%200"> and <img src="https://render.githubusercontent.com/render/math?math=J_%7Bv%7D(t%2C%201)%20%3D%20J_%7Bi%7D(t%2C%201)%20%3D%200">, you have specified 
- <img src="https://render.githubusercontent.com/render/math?math=C_i(t%2C%200)%20%3D%20C_%7Bi%5C_t0%7D"> and <img src="https://render.githubusercontent.com/render/math?math=C_v(t%2C%200)%20%3D%20C_%7Bv%5C_t0%7D">
  which are Dirichlet.
- <img src="https://render.githubusercontent.com/render/math?math=%5Cfrac%7B%5Cpartial%20C_%7BFe%7D(t%2C%201)%7D%7B%5Cpartial%20t%7D%20%3D%20%5Cfrac%7B%5Cpartial%20C_%7BCr%7D(t%2C%201)%7D%7B%5Cpartial%20t%7D%20%3D%200">

  which are equivalent to

  <img src="https://render.githubusercontent.com/render/math?math=C_%7BFe%7D(t%2C%201)%20%3D%20C_%7BFe%5C_t0%7D"> and <img src="https://render.githubusercontent.com/render/math?math=C_%7BCr%7D(t%2C%201)%20%3D%20C_%7BCr%5C_t0%7D">

  which are also Dirichlet.

**Comment by WenjiangHuang111** [comment]
> > The curve I got looks unreasonable to me. Did I mess up something here?
> 
> The equations look OK to me.
> 
> For the boundary conditions, you constrain `C_v` on the left, but `C_i` on the right:
> 
> ```python
> C_v.constrain(C_v_t0, mesh.facesLeft)
> C_i.constrain(C_i_t0, mesh.facesRight)
> ```
> 
> > For BCs, are "Flux of Fe is zero at x=0" and "partial Fe/ partial t =0 at x=1" the same? Cz for 1D, as you said, Jv and div Jv don't appear any difference.
> 
> No. Saying that ![](https://render.githubusercontent.com/render/math?math=%5Cfrac%7B%5Cpartial%20C_%7BFe%7D%7D%7B%5Cpartial%20t%7D%20%3D%200) at x=1 is equivalent to saying that ![](https://render.githubusercontent.com/render/math?math=%5Cnabla%5Ccdot%20J_%7BFe%7D%20%3D%200) at x=1. In other words, the flux of Fe at ![](https://render.githubusercontent.com/render/math?math=x%3D1-%5Cvarepsilon) is the same as at ![](https://render.githubusercontent.com/render/math?math=x%3D1%2B%5Cvarepsilon). It doesn't say anything about the _value_ of the flux, though. Because there are no sources or sinks of Fe, that means that C_Fe is constant.
> 
> Saying that the flux of Fe is zero at x=0 says nothing about how the value of the flux is changing at x=0, thus the divergence can be non-zero, leading to accumulation or depletion at that boundary. This is a Neumann condition. More importantly, no-flux is the natural boundary condition for finite volume and so is addressed automatically by FiPy unless you apply some other constraint.



> You don't have a Neumann condition for Fe and Cr at x=1. In addition to ![](https://render.githubusercontent.com/render/math?math=J_%7BFe%7D(t%2C%200)%20%3D%20J_%7BCr%7D(t%2C%200)%20%3D%200) and ![](https://render.githubusercontent.com/render/math?math=J_%7Bv%7D(t%2C%201)%20%3D%20J_%7Bi%7D(t%2C%201)%20%3D%200), you have specified
> 
> * ![](https://render.githubusercontent.com/render/math?math=C_i(t%2C%200)%20%3D%20C_%7Bi%5C_t0%7D) and ![](https://render.githubusercontent.com/render/math?math=C_v(t%2C%200)%20%3D%20C_%7Bv%5C_t0%7D)
>   which are Dirichlet.
> * ![](https://render.githubusercontent.com/render/math?math=%5Cfrac%7B%5Cpartial%20C_%7BFe%7D(t%2C%201)%7D%7B%5Cpartial%20t%7D%20%3D%20%5Cfrac%7B%5Cpartial%20C_%7BCr%7D(t%2C%201)%7D%7B%5Cpartial%20t%7D%20%3D%200)
>   which are equivalent to
>   ![](https://render.githubusercontent.com/render/math?math=C_%7BFe%7D(t%2C%201)%20%3D%20C_%7BFe%5C_t0%7D) and ![](https://render.githubusercontent.com/render/math?math=C_%7BCr%7D(t%2C%201)%20%3D%20C_%7BCr%5C_t0%7D)
>   which are also Dirichlet.

you are right. I also add the BCs of   ![](https://render.githubusercontent.com/render/math?math=C_%7BFe%7D(t%2C%201)%20%3D%20C_%7BFe%5C_t0%7D) and ![](https://render.githubusercontent.com/render/math?math=C_%7BCr%7D(t%2C%201)%20%3D%20C_%7BCr%5C_t0%7D) at x=1. However, the most weird thing happened in my code is that there is no change on Fe and Cr curve, although I can observe the change of Ci and Cv .
<img width="997" alt="Screen Shot 2020-10-14 at 10 01 25 PM" src="https://user-images.githubusercontent.com/45132014/96067956-cfc34280-0e68-11eb-848e-2138f44a9c17.png">

Is the way I update C_Fe and C_Cr wrong? Because the Eqs and BCs look correct to me. 

WJ

**Comment by guyer** [comment]
> Is the way I update C_Fe and C_Cr wrong? Because the Eqs and BCs look correct to me.

I don't know.

Some things to consider:
- the ranges of values for `C_v` and `C_i` are quite small. It's possible there is a similar range of values for `C_Fe` and `C_Cr` and it's just not visible. You have explicitly set the data range of the `Viewer` to (0, 1) for `C_Fe` and `C_Cr`, but not for `C_v` and `C_i`.
- Your equations are complicated enough that it's not obvious from casual inspection which terms are dominant. `d_Fei0` and `d_Cri0` are more than two orders of magnitude smaller than `d_Fev0` and `d_Crv0`. This could conceivably lead to `C_Fe` and `C_Cr` evolving two orders of magnitude slower than `C_v` and `C_i`. I suggest you examine the full diffusion matrix and the full mobility matrix (of a simple one or two element mesh) to get some insight here.
- Your time step appears to be arbitrary. It's possible you just haven't evolved long enough to see significant changes in all fields.
